simplestore/src/test/java/com/uber/simplestore/impl/SimpleStoreImplTest.java

```diff
@@ -149,15 +149,25 @@ public void deleteAll_openChild() throws Exception {
   }
 
   @Test
-  public void deleteAll_twice() throws Exception {
+  public void deleteAll_twice_sequential_success() throws Exception {
     try (SimpleStore store = SimpleStoreFactory.create(directoryProvider, "twice")) {
       ListenableFuture<Void> first = store.deleteAllNow();
-      ListenableFuture<Void> second = store.deleteAllNow();
       first.get();
+      ListenableFuture<Void> second = store.deleteAllNow();
```

Comment:
I believe we do, and I think the `synchronized` block only protects against interwoven attempts to call `deleteAllNow`, but not enforcing that the stuff in `deleteAllNow` is synchronous e.g. the synchronized block in question can be simplified as:
```
synchronized(lock) {
    setFlush();
    enqueueOnDifferentThread(() -> resetFlush());
}
```
All calls to `setFlush` and `enqueueOnDifferentThread` are guaranteed to happen in that order with no interweaving, but not `setFlush` and `resetFlush`, no? (My knowledge/experience with Java synchronization is empirical so the above paragraph may be lies.)

Also since `deleteAllNow` is a _public_ function, so everything we are doing in this test is technically possible in some downstream code, so I think you are right that the current behavior is at times unpredictable.